lib/pbench/server/api/resources/query_apis/metadata_index/iteration_samples.py

```diff
@@ -0,0 +1,347 @@
+        parameter and instead construct an Elasticsearch query for scrolling
+        based on a client provided scroll id.
+
+        If a scroll_id is specified, the query will return the next page of
+        results of the original query; otherwise we form a new query, either
+        with the specified filters or without any filters (returning up to 10000
+        documents at a time).
+
+        Args:
+            json_data:
+            "run_id": run id of a dataset that client wants to search for
+                    fetching the iteration samples.
+
+            "scroll_id": Server provided Elasticsearch scroll id that client
+                        recieved in the result of the original query with
+                        filters. This can be used to fetch the next page of
+                        the result.
+
+            "filters": key-value representation of query filter parameters to
+                       narrow the search results e.g. {"sample.name": "sample1"}
+
+        EXAMPLE:
+            {
+                "run_id": "1234567"
+                "scroll_id": "cmFuZG9tX3Njcm9sbF9pZF9zdHJpbmdfMg=="
+                "filters": {"sample.name": "sample1"},
+            }
+        """
+        run_id = context["run_id"]
+        dataset = context["dataset"]
+        scroll_id = json_data.get("scroll_id")
+
+        self.logger.info(
+            "Return iteration sample rows {} for dataset {}, prefix {}, run id {}",
+            "next page " if scroll_id else "",
+            dataset,
+            self.prefix,
+            run_id,
+        )
+
+        scroll_id = json_data.get("scroll_id")
+        if scroll_id:
+            # Note: Scroll id expires in 1 minute
+            return {
+                "path": "/_search/scroll",
+                "kwargs": {"json": {"scroll": "1m", "scroll_id": scroll_id}},
+            }
+
+        # Retrieve the ES indices that belong to this run_id
+        index = self.get_index(dataset, "result-data-sample")
+
+        es_filter = [{"match": {"run.id": run_id}}]
+        for filter, value in json_data.get("filters", {}).items():
+            es_filter.append({"match": {filter: value}})
+
+        return {
+            "path": f"/{index}/_search?scroll=1m",
+            "kwargs": {
+                "json": {
+                    "size": IterationSamplesRows.DOCUMENT_SIZE,
+                    "query": {"bool": {"filter": es_filter}},
+                    "sort": [
+                        {"iteration.number": {"order": "asc", "unmapped_type": "long"}}
+                    ],
+                },
+                "params": {"ignore_unavailable": "true"},
+            },
+        }
+
+    def postprocess(self, es_json: JSON, context: CONTEXT) -> Response:
+        """
+        Returns a Flask Response containing a JSON object with keys as
+        results and possibly a scroll_id if the next page of results
+        is available.
```

Comment:
Technically, we don't know if another page of results is available, only that Elasticsearch returned a scroll ID; but I'm not quite sure how to make this clear without weaseling...

Reply:
Actually, I think that ES returns the scroll ID any time you ask for one...regardless of whether there are multiple "pages" or not.  😿

However, we think that we can predict whether there will be another page in all but one case; and, so, when (we think) there is no next page, we won't return a scroll ID (but, in one case, we'll return a scroll ID when the next page is going to be empty...but hopefully the client can handle that).

Reply:
I was commenting on the "is available" ... you're correct that our heuristic isn't 100% reliable, but, more to the point, something like "if we think there might be more data" or "if we can't determine there is no more data".

Reply:
> "if we can't determine there is no more data"

That would be accurate...but I don't know that I like it.  (I don't mind "is available".)